### DISCUSSION TITLE
Recorder stores data inefficiently, timeseries logging probably better?

### DISCUSSION THREAD

**Discussion by oliv3r** [home-assistant/architecture#1100] [General]
Hey all,

I've been looking into the recorder for a whlie, and recently migrated myself to postgresql/timescaledb. Learning more about timeseries, I noticed that currently, the recorder (using postgresql, not sure if sqlite would be any different) does currently store interval based data. E.g. it stores the same value/data multiple times. e.g.
```
homeassistant=# select * from statistics_short_term where metadata_id=22 limit 10;
  id   | created |     created_ts     | metadata_id | start |  start_ts  | mean  |  min  |  max  | last_reset | last_reset_ts | state | sum 
-------+---------+--------------------+-------------+-------+------------+-------+-------+-------+------------+---------------+-------+-----
 31563 |         | 1717294810.4162643 |          22 |       | 1717294500 | 83.36 | 83.36 | 83.36 |            |               |       |    
 31570 |         | 1717295110.4101014 |          22 |       | 1717294800 | 83.36 | 83.36 | 83.36 |            |               |       |    
 31577 |         | 1717295410.4212794 |          22 |       | 1717295100 | 83.36 | 83.36 | 83.36 |            |               |       |    
 31584 |         | 1717295710.4160612 |          22 |       | 1717295400 | 83.36 | 83.36 | 83.36 |            |               |       |    
 31591 |         | 1717296010.4175048 |          22 |       | 1717295700 | 83.36 | 83.36 | 83.36 |            |               |       |    
 31598 |         |  1717296310.416331 |          22 |       | 1717296000 | 83.36 | 83.36 | 83.36 |            |               |       |    
 31605 |         | 1717296610.4163961 |          22 |       | 1717296300 | 83.36 | 83.36 | 83.36 |            |               |       |    
 31612 |         |  1717296910.416549 |          22 |       | 1717296600 | 83.36 | 83.36 | 83.36 |            |               |       |    
 31619 |         | 1717297210.4168398 |          22 |       | 1717296900 | 83.36 | 83.36 | 83.36 |            |               |       |    
 31626 |         | 1717297510.4161253 |          22 |       | 1717297200 | 83.36 | 83.36 | 83.36 |            |               |       |   
```

First thing I noticed here, is that we have the fields `created` and `start` which are of the time `timestamp` however they are not set. Instead, we have a float (`created_ts` and `start_ts` where `last_reset` probably is one as well) representing the unix timestamp in second and nanosecond precision.

So first, why even have this column at all, why not use the actual intended column for this purpose?

Secondly, I'm not sure why we are storing 2 timestamps here, I guess this is 'start of query, end of query'? From a database storage efficiency point of view, this seems a bit wasteful. Granted, this table was intended for 'short term storage' 10 days by default. Still, we could do better, why aren't we? Something to discuss?

Thirdly, the data itself is 'timeseries' in itself. What I've learned for timeseries data, is that you'd actually want to store/commit if/when there was a change on the data. e.g. (from AI :p)
```SQL
INSERT INTO mytable (col1, col2)
SELECT 'data1', 'data2'
WHERE NOT EXISTS (SELECT 1 FROM mytable WHERE col1 = 'data1' AND col2 = 'data2');
```
so something along those lines should be possible.

Fourthly, and this is just my lack of understanding, what is the meaning of 'mean, min max'? If I get a piece of sensor data, it has a single value, so when/how would the min/max/mean value come into play? Could we not do this 'smarter'? For many sensors, we'd only ever get a single value no? Just to enlighten myself, which sensors produce multiple values?
simple example:
```
 3373 |         | 1717383610.2993858 |          22 |       | 1717380000 |             83.36 | 83.36 | 83.36 |            |               |       |    
 3386 |         |  1717387210.297326 |          22 |       | 1717383600 |             83.36 | 83.36 | 83.36 |            |               |       |    
 3399 |         | 1717390810.3020093 |          22 |       | 1717387200 |             83.36 | 83.36 | 83.36 |            |               |       |    
 3412 |         |  1717394410.302218 |          22 |       | 1717390800 |             83.36 | 83.36 | 83.36 |            |               |       |    
 3426 |         | 1717398010.2976599 |          22 |       | 1717394400 |             83.36 | 83.36 | 83.36 |            |               |       |    
 3440 |         | 1717401610.3043294 |          22 |       | 1717398000 |             83.36 | 83.36 | 83.36 |            |               |       |    
 3454 |         | 1717405210.3032625 |          22 |       | 1717401600 | 83.41644472104998 | 83.36 | 83.54 |            |               |       |    
 3468 |         | 1717408810.2744696 |          22 |       | 1717405200 | 83.53999999999999 | 83.54 | 83.54 |            |               |       |    
 3482 |         | 1717412410.3030684 |          22 |       | 1717408800 | 83.53999999999999 | 83.54 | 83.54 |            |               |       |    
 3498 |         | 1717416014.1221318 |          22 |       | 1717412400 | 83.53999999999999 | 83.54 | 83.54 |            |               |       |    
 3530 |         | 1717419610.4555848 |          22 |       | 1717416000 | 83.53999999999999 | 83.54 | 83.54 |            |               |       |    
 3555 |         | 1717423210.4580786 |          22 |       | 1717419600 | 83.53999999999999 | 83.54 | 83.54 |            |               |       |    
```
(ignore the fact that the data is not properly rounded, I think that's a HA bug)

Or is it the case that recorder is doing something smart at the time of transition (3454) where it stores the previous value (in this case as min, but could be min) and the new value (in this case the max)? Is there a (huge?) performance benefit in doing so? Could this not be done in the SQL query itself that retrieves the data? A view/index that could accelerate this? I'm not opposed of this, it just seems like storing a lot of additional data.

Lastly, I've been playing with timescaledb (a postgres extension) lately, and they are tackling the timeseries question exactly for this purpose. Timescale itself claims it's faster and more efficient then influxdb, but in the end, timescale, influx, are all similar solutions to the timeseries problem. Also note, that timescaledb should be fully compatible with the current implementation as it's using standard SQL queries. Also, also note that one cool trick of timescale db, is that it can automatically do aggregation, which is now done manually.

I do get that the recorder is stuck in with legacy reasons and existing databases etc, where migration is not trivial of course. But I also know there's at least some demand for more serious data capture of  HA sensors. The aggregation HA sensors helps a bit with the storage/performance issues, but in the end, it's a whole lot of duplicate data we're storing for no benefit/advantage.

**Comment by frenck** [comment]
Thanks, but... you created an architectural proposal.

Can you describe in detail the technical architectural change you want to make (and how you are going to make it), for which you are seeking architectural approval?

../Frenck

**Reply by oliv3r** [reply]
Hey @frenck 

Firstly, I would start an architectural discussion, leading to a proposal obviously.
For those proposal I'd first need answers to the questions I have asked, but making some assumptions and being ignorant, this is what I'd be suggesting.

Firstly, change the recorder to the following database schema.

```
                                Table "public.statistics_short_term"
    Column     |           Type           | Collation | Nullable |             Default              
---------------+--------------------------+-----------+----------+----------------------------------
 id            | integer                  |           | not null | generated by default as identity
 created       | timestamp with time zone |           |          | 
 metadata_id   | integer                  |           |          | 
 value          | double precision         |           |          | 
```

But obviously this would be a bit shortsighted, as there are probably good reasons for the other values, which I would need help understanding first.

Secondly, I'd propose to, as I wrote above, to only actually store data if there was a change on the data.

```SQL
INSERT INTO statistics_short_term (metadata_id, value)
SELECT '22', '85,4'
WHERE NOT EXISTS (
  SELECT 1
  FROM statistics_short_term
  WHERE metadata_id = '22'
  ORDER BY id DESC
  LIMIT 1
  HAVING value = '85,4'
);
```
(the above query is just an illustration, something from AI, i'd have to spend some time to figure out how to exactly do this, my SQL is a bit rusty)

This would be the first step. But something to get the discussion going? Though again, with the missing insights, there might be a few more subtle steps needed. E.g. the min/max/mean/sum/state/sum thing.

**Reply by frenck** [reply]
Right, but the goal of this repository in general is to discuss proposals, Which this is not; If you want a general discussion and/or gain knowledge on the history of why things are the way they are, the developer channels on Discord or our community forums might be a better place. Of course, git history and all PRs linked to it, will also provide great insights.

../Frenck

**Reply by oliv3r** [reply]
> Right, but the goal of this repository in general is to discuss proposals, Which this is not; If you want a general discussion and/or gain knowledge on the history of why things are the way they are, the developer channels on Discord or our community forums might be a better place. Of course, git history and all PRs linked to it, will also provide great insights.
> 
> ../Frenck

Sure, but I do not use Discord due to privacy reasons (you are the product etc). The forums are of course fair ground, but then this is a 'discussion' section, so it's a bit weird to discus this elsewhere. I would have thought that the discussion section is about discussing things, and an 'issue' would be to make the concrete proposal. But I'll park this for now, and take the discussion over to the forum first. A bit confusing all; but sure.

**Reply by frenck** [reply]
We use the discussion feature of GitHub in this repository to discuss proposals.

See also the pinned message on this repo: https://github.com/home-assistant/architecture/discussions/494

**Reply by oliv3r** [reply]
Indeed, which is what is what I thought I was doing :) 'Propose architectural changes' no? I wanted to avoid spending 3 weeks working out indepth details, to be told 'sorry no'. Best to have a short _discussion_ beforehand. But I clearly missunderstood the intend. Sorry!

**Reply by oliv3r** [reply]
So I have done as you've asked, but maybe in the wrong location? On the forum it is states 'devs don't look here' https://community.home-assistant.io/t/recorder-stores-data-inefficiently-timeseries-logging-probably-better/739377/6

So how to get the discussion started?

**Reply by oliv3r** [reply]
@frenck any thoughts?

**Reply by frenck** [reply]
My thoughts have been expressed above. 

If you want to discuss in general with the community, the forum is the best place. If you have development question, Discord is a better location.

../Frenck
